src/lxc/network.c

```diff
@@ -3529,26 +3525,31 @@ static int lxc_setup_netdev_in_child_namespaces(struct lxc_netdev *netdev)
 				return minus_one_set_errno(-err);
 			}
 		} else {
+			/* Check the gateway address is valid */
+			if (!inet_ntop(AF_INET, netdev->ipv4_gateway, bufinet4, sizeof(bufinet4)))
+				return minus_one_set_errno(-errno);
+
+			/* Try adding a default route to the gateway address */
 			err = lxc_ipv4_gateway_add(netdev->ifindex, netdev->ipv4_gateway);
-			if (err) {
+			if (err < 0) {
+				/* If adding the default route fails, this could be because the
+				 * gateway address is in a different subnet to the container's address.
+				 * To work around this, we try adding a static device route to the
+				 * gateway address first, and then try again.
+				 */
 				err = lxc_ipv4_dest_add(netdev->ifindex, netdev->ipv4_gateway, 32);
-				if (err) {
+				if (err < 0) {
 					errno = -err;
```

Comment:
@brauner in this scenario, I could have remove these `errno = -err` lines and then replaced the `return -1` 2 lines down with `return minus_one_set_errno(-err);`. However that would have resulted in the call to the `SYSERROR` function not having the most recent `errno` set (assuming that the previously called function `lxc_ipv4_dest_add` did not set errno, which is does it some situations and not others).

Alternatively I could have just used `ERROR` so that it wouldnt matter.

Reply:
Meh, your call. No strong opinion from my side.